applications/StructuralMechanicsApplication/custom_elements/base_solid_element.h

```diff
@@ -50,6 +50,68 @@ namespace Kratos
 class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION)  BaseSolidElement
     : public Element
 {
+protected:
+    /**
+     * Internal variables used in the kinematic calculations
+     */
+    struct KinematicVariables
+    {
+        Vector  StrainVector;
+        Vector  StressVector;
+        Vector  N;
+        Matrix  B;
+        double  detF;
+        Matrix  F;
+        double  detJ;
+        double  detJ0;
+        Matrix  J0;
+        Matrix  InvJ0;
+        Matrix  DN_DX;
+        Matrix  D;
+        
+        /**
+         * The variables are initialized
+         * @param StrainSize: The size of the strain vector in Voigt notation
+         * @param Dimension: The size of the strain vector in Voigt notation
+         */
+        void Initialize( 
+            const unsigned int& StrainSize, 
+            const unsigned int& Dimension, 
+            const unsigned int& NumberOfNodes 
+            )
+        {
+            detF = 1.0;
+            detJ0 = 1.0;
+            detJ0 = 1.0;
+            
+            StrainVector.resize(StrainSize, false);
+            noalias(StrainVector) = ZeroVector(StrainSize);
+            
+            StressVector.resize(StrainSize, false);
+            noalias(StressVector) = ZeroVector(StrainSize);
+            
+            N.resize(NumberOfNodes, false);
+            noalias(N) = ZeroVector(NumberOfNodes);
+            
+            B.resize(StrainSize, Dimension * NumberOfNodes, false);
+            noalias(B)  = ZeroMatrix(StrainSize, Dimension * NumberOfNodes);
+            
+            F.resize(Dimension, Dimension,false);
+            noalias(F)  = IdentityMatrix(Dimension);
+            
+            DN_DX.resize(NumberOfNodes, Dimension,false);
+            noalias(DN_DX) = ZeroMatrix(NumberOfNodes, Dimension);
+            
+            D.resize(StrainSize, StrainSize,false);
+            noalias(D) = ZeroMatrix(StrainSize, StrainSize);
```

Comment:
@RiccardoRossi The struct, and B in  particular, is built just one time outside the integration loop, during the loop is just updated. It is basically the same thing than before, but in a struct, so we can move things to the base class and easily implement the computation in the integration points